src/value/borrowed.rs

```diff
@@ -0,0 +1,229 @@
+}
+
+impl<'val> BorrowedStruct<'val> {
+    pub fn new(fields: Vec<(BorrowedSymbolToken<'val>, BorrowedElement<'val>)>) -> Self {
+        Self { fields }
+    }
+}
+
+impl<'val> Struct for BorrowedStruct<'val> {
+    type FieldName = BorrowedSymbolToken<'val>;
+    type Element = BorrowedElement<'val>;
+
+    fn iter<'a>(
+        &'a self,
+    ) -> Box<dyn Iterator<Item = (&'a Self::FieldName, &'a Self::Element)> + 'a> {
+        // convert &(k, v) -> (&k, &v)
+        Box::new(self.fields.iter().map(|kv| match &kv {
+            (k, v) => (k, v),
+        }))
+    }
+}
+
+// TODO replace the references with `Cow` and bridge to the owned APIs for mutability
+
+/// Variants for all borrowed version _values_ within an [`Element`].
+#[derive(Debug, Clone)]
+pub enum BorrowedValue<'val> {
+    Null(IonType),
+    Integer(AnyInt),
+    String(&'val str),
+    Symbol(BorrowedSymbolToken<'val>),
+    SExpression(BorrowedSequence<'val>),
+    List(BorrowedSequence<'val>),
+    Struct(BorrowedStruct<'val>),
+    // TODO fill this in with the rest...
+}
+
+/// A borrowed implementation of [`Element`]
+#[derive(Debug, Clone)]
+pub struct BorrowedElement<'val> {
+    annotations: Vec<BorrowedSymbolToken<'val>>,
+    value: BorrowedValue<'val>,
+}
+
+impl<'val> BorrowedElement<'val> {
+    pub fn new(annotations: Vec<BorrowedSymbolToken<'val>>, value: BorrowedValue<'val>) -> Self {
+        Self { annotations, value }
+    }
+}
+
+impl<'val> From<BorrowedValue<'val>> for BorrowedElement<'val> {
+    /// Constructs a [`BorrowedElement`] without annotations from this value.
+    fn from(val: BorrowedValue<'val>) -> Self {
+        Self::new(vec![], val)
+    }
+}
+
+impl<'val> Element for BorrowedElement<'val> {
+    type SymbolToken = BorrowedSymbolToken<'val>;
+    type Sequence = BorrowedSequence<'val>;
+    type Struct = BorrowedStruct<'val>;
+
+    fn ion_type(&self) -> IonType {
+        use BorrowedValue::*;
+        match &self.value {
+            Null(t) => *t,
+            Integer(_) => IonType::Integer,
+            String(_) => IonType::String,
+            Symbol(_) => IonType::Symbol,
+            SExpression(_) => IonType::SExpression,
+            List(_) => IonType::List,
+            Struct(_) => IonType::Struct,
+        }
+    }
+
+    fn annotations<'a>(&'a self) -> Box<dyn Iterator<Item = &'a Self::SymbolToken> + 'a> {
+        Box::new(self.annotations.iter())
+    }
+
+    fn is_null(&self) -> bool {
+        match &self.value {
+            BorrowedValue::Null(_) => true,
+            _ => false,
+        }
+    }
+
+    fn as_any_int(&self) -> Option<&AnyInt> {
+        match &self.value {
+            BorrowedValue::Integer(i) => Some(i),
+            _ => None,
+        }
+    }
+
+    fn as_str(&self) -> Option<&str> {
+        match &self.value {
+            BorrowedValue::String(text) => Some(*text),
+            BorrowedValue::Symbol(sym) => sym.text(),
+            _ => None,
+        }
+    }
```

Comment:
I'm not sure if this is a DOM-like API pattern (if so, FFTI), but I noticed it in the Kotlin IonElement, too: is there a reason to have these `as_*()` methods in the trait signature?

Especially in Rust, it seems like `impl From` would be helpful here, too. Then callers can specify `let f: Whatever = foo.into()` without having to call `let f = foo.as_whatever()`--and `Into<Whatever>` if they want to parameterize their own types. In other words, it seems like it would let `rustc` do more of the work.

Reply:
Yes, this PR doesn't have all the `From`/`Into` conversion operators nor does it deal with `ToOwned` traits either.  Those traits could be defined outside of the trait as potentially blanket implementations of the above traits.